README.md
# Kepler
This package provides an api for encrypting data using public and private 
keys created by the secp256k1 elliptic curve. 

This is a fork of [secp256k1Cipher](https://github.com/ipconfiger/secp256k1Cipher).

Example usage.
```dart
 
  // Create Alice's keypair
  var alice = generateKeyPair();

  print(
    "alice private key: " +
        strinifyPrivateKey(alice.privateKey as ECPrivateKey),
  );
  print(
    "alice public key: " + strinifyPublicKey(alice.publicKey as ECPublicKey),
  );

  // Create Bob's keypair
  var bob = generateKeyPair();

  // This is what alice wants to say to bob
  var rawStr = 'Encrypt and decrypt data using secp256k1';

  // use alic's privatekey and bob's publickey means alice says to bob
  var encMap = pubkeyEncrypt(
    strinifyPrivateKey(alice.privateKey as ECPrivateKey),
    strinifyPublicKey(bob.publicKey as ECPublicKey),
    rawStr,
  );

  // Get encrypted base64 string
  var encStr = encMap['enc'];
  print("encrypted text: " + encStr);

  // Get random IV
  var iv = encMap['iv'];
  print("iv: " + iv);
  
  // Now, you can send enc_str and IV to Bob

  // Use bob's privatekey and alice's publickey to decrypt alices message, for Bob to read.
  var decryptd = privateDecrypt(
    strinifyPrivateKey(bob.privateKey as ECPrivateKey),
    strinifyPublicKey(alice.publicKey as ECPublicKey),
    encStr,
    iv,
  );
  print('decrypted text: $decryptd');
```

CHANGELOG.md
Changelog
=========

### Version 1.0.3 (2022-06-11)
* 1. Bumped packages to latest
* 2. Added strict linting rules 

### Version 1.0.2 (2022-01-25)
* 1. Changed functions to be static methods of class Kepler
* 2. Updated pointycastle version to latest

### Version 1.0.1 (2021-08-04)
* 1. Switched flutter_test dev dependency for test dev dependency
* 2. Updated pointycastle version to latest


### Version 1.0.0 (2021-07-10)
* 1. Initial release of forked repo.
* 2. Migrated codebase to null safety.
* 3. Fixed various linting errors and variable names.
* 4. Added the ability to seed the key generator.


test/impl_test.dart
// ignore_for_file: avoid_print

library kepler.test.impl_test;

import 'dart:typed_data';
import 'dart:convert' as convert;
import 'package:base58check/base58.dart';
import 'package:hex/hex.dart';
import 'package:kepler/src/kepler.dart';
import 'package:pointycastle/digests/ripemd160.dart';
import 'package:test/test.dart';
import 'package:kepler/kepler.dart';
import "package:pointycastle/pointycastle.dart";

void main() {
  group('Keys', () {
    test("genaddr", () {
      Digest sha256 = Digest("SHA-256");
      Digest ripemd = RIPEMD160Digest();
      final pubkey = Kepler.loadPublicKey(
          '50863AD64A87AE8A2FE83C1AF1A8403CB53F53E486D8511DAD8A04887E5B23522CD470243453A299FA9E77237716103ABC11A1DF38855ED6F2EE187E9C582BA6');
      final pubBytes = pubkey.Q!.getEncoded(false);
      final shaHash = sha256.process(pubBytes);
      final ripHash = ripemd.process(shaHash);
      final hexHash = HEX.encode(ripHash.toList());
      print(hexHash);

      final networkHash = [0x00] + ripHash.toList();
      final check1 = sha256.process(Uint8List.fromList(networkHash));
      final check2 = sha256.process(check1);
      final finalCheck = check2.sublist(0, 4);
      final codeList = networkHash + finalCheck;
      print(HEX.encode(codeList));
      Base58Encoder b58 = const Base58Encoder(
          '123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz');
      final b58Str = b58.convert(codeList);
      print(b58Str);
    });

    test('Generate Keys', () {
      final lines = [];
      for (var _idx = 0; _idx < 100; _idx++) {
        final keypare = Kepler.generateKeyPair();
        final ECPublicKey pubkey = keypare.publicKey as ECPublicKey;
        final ECPrivateKey prikey = keypare.privateKey as ECPrivateKey;
        final line = [
          Kepler.strinifyPublicKey(pubkey),
          Kepler.strinifyPrivateKey(prikey),
          pubkey.Q!.x!.toBigInteger()!.toRadixString(16),
          pubkey.Q!.y!.toBigInteger()!.toRadixString(16)
        ];
        final row = line.join(',');
        lines.add(row);
      }
      final txt = lines.join('\n');
      print(txt);
      expect(true, equals(true));
    });
    test('Save and restore private key', () {
      const localPrivate =
          'eaa692953a60ff85beecdf9647807f5e1bd665aa342c3c1d893b54bccf816ff5';
      const remotePublic =
          '02766171786852c788bfac4622b302b1c42ca77e3bfdabc56454a4ca5647ac4eba';
      const enc = 'd5dTsgku25ylogZ7Yjs=';
      const iv = 'MRz0cLx8QL4=';
      final raw = Kepler.privateDecrypt(localPrivate, remotePublic, enc, iv);
      print('raw: $raw');
    });
    test('Save and restore public key', () async {
      for (var _idx = 0; _idx < 100; _idx++) {
        const alicPubkey =
            '3d6b2142489ffa6d221da41e75e6c08a44a8d0e682b9fa6d768594d94da2adeeb85e248fa05dedcc4f95c32ab8707bb0ba579fd4b41bf28a0df5bfd7f731b809';
        const alicPrikey =
            '462c58255c68a0a1c1b5c89baa99688c81760169ed8c1502d53e50a820aed90a';
        final bob = Kepler.generateKeyPair();
        print('success idx: $_idx');
        final s1 = Kepler.rawSecret(
            alicPrikey, Kepler.strinifyPublicKey(bob.publicKey as ECPublicKey));
        final s2 = Kepler.rawSecret(
          Kepler.strinifyPrivateKey(bob.privateKey as ECPrivateKey),
          alicPubkey,
        );
        expect(s1, equals(s2));
      }
    });
    test('Make request', () async {
      const convert.Utf8Encoder encoder = convert.Utf8Encoder();
      const remotePubkey =
          '02be8d8a7b5056de7a7074236100d094ebe86cce33d62469956203022af1f3e556';
      final myKP = Kepler.generateKeyPair();
      const data = 'abcdefg';
      final strPrivKey =
          Kepler.strinifyPrivateKey(myKP.privateKey as ECPrivateKey);
      final strPubKey = Kepler.strinifyPublicKey(myKP.publicKey as ECPublicKey);
      final enced = Kepler.pubkeyEncryptRaw(
          strPrivKey, remotePubkey, Uint8List.fromList(encoder.convert(data)));
      final List<int> dataArr = [];
      dataArr.addAll(encoder.convert(strPubKey));
      dataArr.addAll(enced['enc'].toList());
      print(convert.base64.encode(dataArr));
      expect(true, true);
    });
    test('Test Decrypt', () async {
      const convert.Utf8Decoder decoder = convert.Utf8Decoder();
      const myPrivate =
          '1241ae561074f703c259da27036af3510640bbd6a79ceed7eaea4b3b566befe9';
      const message =
          'MDNjMThhN2RlN2I3ZjQwYTgwMDQwMDg1OGUyMTIwNmYyNzdiYjJhZGMwZjAyMDUzYjMzODYyZDgwY2Q0M2YxN2JhqDMwhGPjj2d4hpz2hfjjyRHQ';
      final rawData = convert.base64.decode(message);
      final pubKey = decoder.convert(rawData.getRange(0, 66).toList());
      final payload = rawData.getRange(66, rawData.length).toList();
      final decrypted = Kepler.privateDecryptRaw(
        myPrivate,
        pubKey,
        Uint8List.fromList(payload),
      );
      print("raw message=${decoder.convert(decrypted)}");
      expect(true, true);
    });

    test('Sign and Verify', () {
      final alice = Kepler.generateKeyPair();
      const message = "Mary has a little sheep";
      final alicePubkey =
          Kepler.strinifyPublicKey(alice.publicKey as ECPublicKey);
      final alicePrivatekey =
          Kepler.strinifyPrivateKey(alice.privateKey as ECPrivateKey);
      final signature = Kepler.privateSign(alicePrivatekey, message);
      expect(true, Kepler.publicVerify(alicePubkey, message, signature));
    });

    test('Encrypt and Decrypt', () {
      int microSeconds = 0;
      for (var i = 0; i < 10; i++) {
        const alicPubkey =
            '5cb38e0c76f2b28e112e78d96d46e79b04585f17c3bb81a11ad3ad327d9ccaf815b0d2c770fd31c7224671378d7129cdd3dba97ca1efd016e2a580048c6eec46';
        const alicPrikey =
            '9717f155a64b67e5aa22a9552824237119a373b84ffe62eb435cac6581099767';
        var bob = Kepler.generateKeyPair();
        var rawStr = 'Very secret stuff in this string of text';
        final t1 = DateTime.now().millisecondsSinceEpoch;
        var encMap = Kepler.pubkeyEncrypt(alicPrikey,
            Kepler.strinifyPublicKey(bob.publicKey as ECPublicKey), rawStr);
        microSeconds += (DateTime.now().millisecondsSinceEpoch - t1);
        var encStr = encMap['enc'];
        var iv = encMap['iv'];
        var decryptd = Kepler.privateDecrypt(
          Kepler.strinifyPrivateKey(bob.privateKey as ECPrivateKey),
          alicPubkey,
          encStr!,
          iv!,
        );
        expect(decryptd, equals(rawStr));
      }
      print('avg: ${microSeconds / 100} ms');
    });
  });
}


lib/src/kepler.dart
// ignore_for_file: implementation_imports

import "dart:typed_data";
import "dart:math";
import 'dart:convert' as convert;
import "package:pointycastle/export.dart";
import 'package:hex/hex.dart';
import 'package:base58check/base58.dart';
import 'operator.dart';
import 'package:pointycastle/src/impl/secure_random_base.dart';
import "package:pointycastle/src/registry/registry.dart";
import "package:pointycastle/src/ufixnum.dart";

class NullSecureRandom extends SecureRandomBase {
  static final FactoryConfig factoryConfig =
      StaticFactoryConfig(SecureRandom, "Null", () => NullSecureRandom());

  var _nextValue = 0;

  @override
  String get algorithmName => "Null";

  @override
  void seed(CipherParameters params) {}

  @override
  int nextUint8() => clip8(_nextValue++);
}

class Kepler {
  static final SHA3Digest sha3digest = SHA3Digest(256 ~/ 8 * 8);

  /// return a hex string version privateKey
  static String strinifyPrivateKey(ECPrivateKey privateKey) {
    assert(privateKey.d != null);
    return privateKey.d!.toRadixString(16);
  }

  static String leftPadding(String s, int width) {
    const paddingData = '000000000000000';
    final paddingWidth = width - s.length;
    if (paddingWidth < 1) {
      return s;
    }
    return "${paddingData.substring(0, paddingWidth)}$s";
  }

  /// return a BTC Address
  static String btcAddress(ECPublicKey pubkey) {
    assert(pubkey.Q != null);
    Digest sha256 = Digest("SHA-256");
    Digest ripemd = RIPEMD160Digest();
    final pubBytes = pubkey.Q!.getEncoded(false);
    final shaHash = sha256.process(pubBytes);
    final ripHash = ripemd.process(shaHash);
    final networkHash = [0x00] + ripHash.toList();
    final check1 = sha256.process(Uint8List.fromList(networkHash));
    final check2 = sha256.process(check1);
    final finalCheck = check2.sublist(0, 4);
    final codeList = networkHash + finalCheck;
    Base58Encoder b58 = const Base58Encoder(
      '123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz',
    );
    return b58.convert(codeList);
  }

  /// return a ETH Address
  static String ethAddress(ECPublicKey pubkey) {
    assert(pubkey.Q != null);
    sha3digest.reset();
    final pubBytes = pubkey.Q!.getEncoded(false);
    final addressBytes = sha3digest.process(pubBytes);
    final hexString = HEX.encode(addressBytes);
    return '0x${hexString.substring(24)}';
  }

  /// return a hex string version publicKey
  static String strinifyPublicKey(ECPublicKey publicKey) {
    assert(publicKey.Q != null);
    Uint8List compressedKey = publicKey.Q!.getEncoded(true);
    final codeList = compressedKey.toList();
    return codeList.map((w) {
      final hx = w.toRadixString(16);
      if (hx.length < 2) {
        return '0$hx';
      }
      return hx;
    }).join('');
    //print('bytes:${raw_bytes}');
    //var x_str = left_padding(publicKey.Q.x.toBigInteger().toRadixString(16), 64);
    //var y_str = left_padding(publicKey.Q.y.toBigInteger().toRadixString(16), 64);
    //return "${x_str}${y_str}";
  }

  static String privateSign(String strPrivateKey, String message) {
    ECPrivateKey privateKey = loadPrivateKey(strPrivateKey);
    ECDSASigner singer = ECDSASigner(SHA512Digest(), Mac('SHA-512/HMAC'));
    var privParams =
        PrivateKeyParameter(ECPrivateKey(privateKey.d, privateKey.parameters));
    ParametersWithRandom signParams =
        ParametersWithRandom(privParams, NullSecureRandom());
    singer.init(true, signParams);
    ECSignature signature = singer.generateSignature(
        Uint8List.fromList(convert.utf8.encode(message))) as ECSignature;
    final xs = signature.r.toRadixString(16);
    final ys = signature.s.toRadixString(16);
    final hexX = leftPadding(xs, 64);
    final hexY = leftPadding(ys, 64);
    return hexX + hexY;
  }

  static bool publicVerify(
    String strPublicKey,
    String message,
    String strSignature,
  ) {
    ECPublicKey publicKey = loadPublicKey(strPublicKey);
    ECDSASigner verifySinger = ECDSASigner(SHA512Digest(), Mac('SHA-512/HMAC'));
    var pubkeyParam =
        PublicKeyParameter(ECPublicKey(publicKey.Q, publicKey.parameters));

    final strR = strSignature.substring(0, 64);
    final strS = strSignature.substring(64, 128);
    final r = BigInt.parse(strR, radix: 16);
    final s = BigInt.parse(strS, radix: 16);

    ECSignature signature = ECSignature(r, s);
    verifySinger.init(false, pubkeyParam);
    return verifySinger.verifySignature(
        Uint8List.fromList(convert.utf8.encode(message)), signature);
  }

  /// return a privateKey from hex string
  static ECPrivateKey loadPrivateKey(String storedkey) {
    final d = BigInt.parse(storedkey, radix: 16);
    final param = ECCurve_secp256k1();
    return ECPrivateKey(d, param);
  }

  /// return a publicKey from hex string
  static ECPublicKey loadPublicKey(String storedkey) {
    final param = ECCurve_secp256k1();
    if (storedkey.length < 120) {
      List<int> codeList = [];
      for (var _idx = 0; _idx < storedkey.length - 1; _idx += 2) {
        final hexStr = storedkey.substring(_idx, _idx + 2);
        codeList.add(int.parse(hexStr, radix: 16));
      }
      final Q = param.curve.decodePoint(codeList);
      return ECPublicKey(Q, param);
    } else {
      final x = BigInt.parse(storedkey.substring(0, 64), radix: 16);
      final y = BigInt.parse(storedkey.substring(64), radix: 16);
      final Q = param.curve.createPoint(x, y);
      return ECPublicKey(Q, param);
    }
  }

  /// return a ECPoint data secret
  static ECPoint rawSecret(String privateString, String publicString) {
    final privateKey = loadPrivateKey(privateString);
    final publicKey = loadPublicKey(publicString);
    assert(privateKey.d != null && publicKey.Q != null);
    final secret = scalarMultiple(
      privateKey.d!,
      publicKey.Q!,
    );
    return secret;
  }

  /// return a Bytes data secret
  static List<List<int>> byteSecret(String privateString, String publicString) {
    final secret = rawSecret(privateString, publicString);
    assert(secret.x != null && secret.y != null);
    final xs = secret.x!.toBigInteger()!.toRadixString(16);
    final ys = secret.y!.toBigInteger()!.toRadixString(16);
    final hexX = leftPadding(xs, 64);
    final hexY = leftPadding(ys, 64);
    final secretBytes = Uint8List.fromList(HEX.decode('$hexX$hexY'));
    final pair = [
      secretBytes.sublist(0, 32),
      secretBytes.sublist(32, 40),
    ];
    return pair;
  }

  /// Encrypt data using target public key
  static Map<String, String> pubkeyEncrypt(
    String privateString,
    String publicString,
    String message,
  ) {
    convert.Utf8Encoder encoder = const convert.Utf8Encoder();
    final enced = pubkeyEncryptRaw(privateString, publicString,
        Uint8List.fromList(encoder.convert(message)));
    return {'enc': convert.base64.encode(enced['enc']), 'iv': enced['iv']};
  }

  static Map<String, dynamic> pubkeyEncryptRaw(
    String privateString,
    String publicString,
    Uint8List data,
  ) {
    final secretIV = byteSecret(privateString, publicString);
    final secret = Uint8List.fromList(secretIV[0]);
    final iv = Uint8List.fromList(secretIV[1]);
    Salsa20Engine _cipher = Salsa20Engine();
    _cipher.reset();
    _cipher.init(true, _buildParams(secret, iv));
    final Uint8List encData = _cipher.process(data);
    return {'enc': encData, 'iv': convert.base64.encode(iv)};
  }

  /// Decrypt data using self private key
  static String privateDecrypt(
      String privateString, String publicString, String b64encoded,
      [String b64IV = ""]) {
    Uint8List encdData = convert.base64.decode(b64encoded);
    final rawData =
        privateDecryptRaw(privateString, publicString, encdData, b64IV);
    convert.Utf8Decoder decode = const convert.Utf8Decoder();
    return decode.convert(rawData.toList());
  }

  static Uint8List privateDecryptRaw(
      String privateString, String publicString, Uint8List encdData,
      [String b64IV = ""]) {
    final secretIV = byteSecret(privateString, publicString);
    final secret = Uint8List.fromList(secretIV[0]);
    final iv = b64IV.length > 6
        ? convert.base64.decode(b64IV)
        : Uint8List.fromList(secretIV[1]);
    Salsa20Engine _cipher = Salsa20Engine();
    _cipher.reset();
    _cipher.init(false, _buildParams(secret, iv));
    return _cipher.process(encdData);
  }

  static ParametersWithIV<KeyParameter> _buildParams(
      Uint8List key, Uint8List iv) {
    return ParametersWithIV<KeyParameter>(KeyParameter(key), iv);
  }

  /// Generate Keypair
  static AsymmetricKeyPair<PublicKey, PrivateKey> generateKeyPair({
    Uint8List? seed,
  }) {
    var keyParams = ECCurve_secp256k1();
    var random = FortunaRandom();
    if (seed == null) {
      random.seed(KeyParameter(_seed(32)));
    } else {
      random.seed(KeyParameter(seed));
    }
    BigInt n = keyParams.n;
    int nBitLength = n.bitLength;
    BigInt d;
    do {
      d = random.nextBigInteger(nBitLength);
    } while (d == BigInt.zero || (d >= n));
    ECPoint q = scalarMultiple(d, keyParams.G);
    return AsymmetricKeyPair(
      ECPublicKey(q, keyParams),
      ECPrivateKey(d, keyParams),
    );
  }

  static Uint8List _seed(length) {
    var random = Random.secure();
    var seed = List<int>.generate(length, (_) => random.nextInt(256));
    return Uint8List.fromList(seed);
  }
}


example/lib/main.dart
// ignore_for_file: avoid_print

import 'package:pointycastle/pointycastle.dart';
import 'package:kepler/kepler.dart';

void main() {
  // Create Alice's keypair

  var alice = Kepler.generateKeyPair();

  print(
    "alice private key: " +
        Kepler.strinifyPrivateKey(alice.privateKey as ECPrivateKey),
  );
  print(
    "alice public key: " +
        Kepler.strinifyPublicKey(alice.publicKey as ECPublicKey),
  );

  // Create Bob's keypair
  var bob = Kepler.generateKeyPair();

  // This is what alice wants to say to bob
  var rawStr = 'Encrypt and decrypt data using secp256k1';

  // use alic's privatekey and bob's publickey means alice says to bob
  var encMap = Kepler.pubkeyEncrypt(
    Kepler.strinifyPrivateKey(alice.privateKey as ECPrivateKey),
    Kepler.strinifyPublicKey(bob.publicKey as ECPublicKey),
    rawStr,
  );

  // Get encrypted base64 string
  var encStr = encMap['enc'];
  print("encrypted text: " + encStr!);

  // Get random IV
  var iv = encMap['iv'];
  print("iv: " + iv!);
  // Now, you can send enc_str and IV to Bob

  // Use bob's privatekey and alice's publickey to decrypt alices message, for Bob to read.
  var decryptd = Kepler.privateDecrypt(
    Kepler.strinifyPrivateKey(bob.privateKey as ECPrivateKey),
    Kepler.strinifyPublicKey(alice.publicKey as ECPublicKey),
    encStr,
    iv,
  );
  print('decrypted text: $decryptd');
}


lib/src/operator.dart
import 'package:pointycastle/export.dart';

BigInt theP = BigInt.parse(
    "fffffffffffffffffffffffffffffffffffffffffffffffffffffffefffffc2f",
    radix: 16);
BigInt theN = BigInt.parse(
    "fffffffffffffffffffffffffffffffebaaedce6af48a03bbfd25e8cd0364141",
    radix: 16);

bool isOnCurve(ECPoint point) {
  assert(point.x != null &&
      point.y != null &&
      point.curve.a != null &&
      point.curve.b != null);
  final x = point.x!.toBigInteger();
  final y = point.y!.toBigInteger();
  final rs = (y! * y -
          x! * x * x -
          point.curve.a!.toBigInteger()! * x -
          point.curve.b!.toBigInteger()!) %
      theP;
  return rs == BigInt.from(0);
}

BigInt inverseMod(BigInt k, BigInt p) {
  if (k.compareTo(BigInt.zero) == 0) {
    throw IntegerDivisionByZeroException();
  }
  if (k < BigInt.from(0)) {
    return p - inverseMod(-k, p);
  }
  var s = [BigInt.from(0), BigInt.from(1), BigInt.from(1)];
  var t = [BigInt.from(1), BigInt.from(0), BigInt.from(0)];
  var r = [p, k, k];
  while (r[0] != BigInt.from(0)) {
    var quotient = r[2] ~/ r[0];
    r[1] = r[2] - quotient * r[0];
    r[2] = r[0];
    r[0] = r[1];
    s[1] = s[2] - quotient * s[0];
    s[2] = s[0];
    s[0] = s[1];
    t[1] = t[2] - quotient * t[0];
    t[2] = t[0];
    t[0] = t[1];
  }
  final gcd = r[2];
  final x = s[2];
  // final y = t[2];
  assert(gcd == BigInt.from(1));
  assert((k * x) % p == BigInt.from(1));
  return x % p;
}

ECPoint pointNeg(ECPoint point) {
  assert(isOnCurve(point));
  assert(point.x != null || point.y != null);
  final x = point.x!.toBigInteger();
  final y = point.y!.toBigInteger();
  final result = point.curve.createPoint(x!, -y! % theP);
  assert(isOnCurve(result));
  return result;
}

ECPoint pointAdd(ECPoint? point1, ECPoint? point2) {
  if (point1 == null) {
    return point2!;
  }
  if (point2 == null) {
    return point1;
  }
  assert(isOnCurve(point1));
  assert(isOnCurve(point2));
  final x1 = point1.x!.toBigInteger();
  final y1 = point1.y!.toBigInteger();
  final x2 = point2.x!.toBigInteger();
  final y2 = point2.y!.toBigInteger();

  // assert(x1 != x2 && y1 == y2);
  // if (x1 == x2 && y1 != y2) {
  //   return null;
  // }
  var m;
  if (x1 == x2) {
    m = (BigInt.from(3) * x1! * x1 + point1.curve.a!.toBigInteger()!) *
        inverseMod(BigInt.from(2) * y1!, theP);
  } else {
    m = (y1! - y2!) * inverseMod(x1! - x2!, theP);
  }
  final x3 = m * m - x1 - x2;
  final y3 = y1 + m * (x3 - x1);
  ECPoint result = point1.curve.createPoint(x3 % theP, -y3 % theP);
  assert(isOnCurve(result));
  return result;
}

ECPoint scalarMultiple(BigInt k, ECPoint point) {
  assert(isOnCurve(point));
  assert((k % theN).compareTo(BigInt.zero) != 0);
  assert(point.x != null && point.y != null);
  if (k < BigInt.from(0)) {
    return scalarMultiple(-k, pointNeg(point));
  }
  ECPoint? result;
  ECPoint addend = point;
  while (k > BigInt.from(0)) {
    if (k & BigInt.from(1) > BigInt.from(0)) {
      result = pointAdd(result, addend);
    }
    addend = pointAdd(addend, addend);
    k >>= 1;
  }
  assert(isOnCurve(result!));
  return result!;
}


lib/src/secp256k1Cipher.dart
import "dart:typed_data";
import "dart:math";
import 'dart:convert' as convert;
import 'package:pointycastle/digests/sha3.dart';
import "package:pointycastle/pointycastle.dart";
import "package:pointycastle/export.dart";
import "package:pointycastle/api.dart";
import "package:pointycastle/ecc/api.dart";
import "package:pointycastle/ecc/curves/secp256k1.dart";
import "package:pointycastle/random/fortuna_random.dart";
import 'package:pointycastle/stream/salsa20.dart';
import 'package:hex/hex.dart';
import 'package:base58check/base58.dart';
import 'package:pointycastle/digests/ripemd160.dart';
import 'operator.dart';
import 'package:pointycastle/src/impl/secure_random_base.dart';
import "package:pointycastle/src/registry/registry.dart";
import "package:pointycastle/src/ufixnum.dart";

class NullSecureRandom extends SecureRandomBase {
  static final FactoryConfig factoryConfig =
      StaticFactoryConfig(SecureRandom, "Null", () => NullSecureRandom());

  var _nextValue = 0;

  String get algorithmName => "Null";

  void seed(CipherParameters params) {}

  int nextUint8() => clip8(_nextValue++);
}

/// return a hex string version privateKey
String strinifyPrivateKey(ECPrivateKey privateKey) {
  assert(privateKey.d != null);
  return privateKey.d!.toRadixString(16);
}

String leftPadding(String s, int width) {
  final paddingData = '000000000000000';
  final paddingWidth = width - s.length;
  if (paddingWidth < 1) {
    return s;
  }
  return "${paddingData.substring(0, paddingWidth)}$s";
}

/// return a BTC Address
String btcAddress(ECPublicKey pubkey) {
  assert(pubkey.Q != null);
  Digest sha256 = Digest("SHA-256");
  Digest ripemd = RIPEMD160Digest();
  final pubBytes = pubkey.Q!.getEncoded(false);
  final shaHash = sha256.process(pubBytes);
  final ripHash = ripemd.process(shaHash);
  // 生成验证
  final networkHash = [0x00] + ripHash.toList();
  final check1 = sha256.process(Uint8List.fromList(networkHash));
  final check2 = sha256.process(check1);
  final finalCheck = check2.sublist(0, 4);
  final codeList = networkHash + finalCheck;
  Base58Encoder b58 = Base58Encoder(
    '123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz',
  );
  return b58.convert(codeList);
}

const int _shaBytes = 256 ~/ 8;
final SHA3Digest sha3digest = SHA3Digest(_shaBytes * 8);

String ethAddress(ECPublicKey pubkey) {
  assert(pubkey.Q != null);
  sha3digest.reset();
  final pubBytes = pubkey.Q!.getEncoded(false);
  final addressBytes = sha3digest.process(pubBytes);
  final hexString = HEX.encode(addressBytes);
  return '0x${hexString.substring(24)}';
}

/// return a hex string version publicKey
String strinifyPublicKey(ECPublicKey publicKey) {
  assert(publicKey.Q != null);
  Uint8List compressedKey = publicKey.Q!.getEncoded(true);
  final codeList = compressedKey.toList();
  //print('raw codes:${code_list}');
  return codeList.map((w) {
    final hx = w.toRadixString(16);
    if (hx.length < 2) {
      return '0$hx';
    }
    return hx;
  }).join('');
  //print('bytes:${raw_bytes}');
  //var x_str = left_padding(publicKey.Q.x.toBigInteger().toRadixString(16), 64);
  //var y_str = left_padding(publicKey.Q.y.toBigInteger().toRadixString(16), 64);
  //return "${x_str}${y_str}";
}

String privateSign(String strPrivateKey, String message) {
  ECPrivateKey privateKey = loadPrivateKey(strPrivateKey);
  ECDSASigner singer = ECDSASigner(SHA512Digest(), Mac('SHA-512/HMAC'));
  var privParams =
      PrivateKeyParameter(ECPrivateKey(privateKey.d, privateKey.parameters));
  var signParams = () => ParametersWithRandom(privParams, NullSecureRandom());
  singer.init(true, signParams());
  ECSignature signature =
      singer.generateSignature(Uint8List.fromList(convert.utf8.encode(message)))
          as ECSignature;
  final xs = signature.r.toRadixString(16);
  final ys = signature.s.toRadixString(16);
  final hexX = leftPadding(xs, 64);
  final hexY = leftPadding(ys, 64);
  return hexX + hexY;
}

bool publicVerify(String strPublicKey, String message, String strSignature) {
  ECPublicKey publicKey = loadPublicKey(strPublicKey);
  ECDSASigner verifySinger = ECDSASigner(SHA512Digest(), Mac('SHA-512/HMAC'));
  var pubkeyParam =
      PublicKeyParameter(ECPublicKey(publicKey.Q, publicKey.parameters));

  final strR = strSignature.substring(0, 64);
  final strS = strSignature.substring(64, 128);
  final r = BigInt.parse(strR, radix: 16);
  final s = BigInt.parse(strS, radix: 16);

  ECSignature signature = ECSignature(r, s);
  verifySinger.init(false, pubkeyParam);
  return verifySinger.verifySignature(
      Uint8List.fromList(convert.utf8.encode(message)), signature);
}

/// return a privateKey from hex string
ECPrivateKey loadPrivateKey(String storedkey) {
  final d = BigInt.parse(storedkey, radix: 16);
  final param = ECCurve_secp256k1();
  return ECPrivateKey(d, param);
}

/// return a publicKey from hex string
ECPublicKey loadPublicKey(String storedkey) {
  final param = ECCurve_secp256k1();
  if (storedkey.length < 120) {
    List<int> codeList = [];
    for (var _idx = 0; _idx < storedkey.length - 1; _idx += 2) {
      final hexStr = storedkey.substring(_idx, _idx + 2);
      codeList.add(int.parse(hexStr, radix: 16));
    }
    final Q = param.curve.decodePoint(codeList);
    return ECPublicKey(Q, param);
  } else {
    final x = BigInt.parse(storedkey.substring(0, 64), radix: 16);
    final y = BigInt.parse(storedkey.substring(64), radix: 16);
    final Q = param.curve.createPoint(x, y);
    return ECPublicKey(Q, param);
  }
}

/// return a ECPoint data secret
ECPoint rawSecret(String privateString, String publicString) {
  final privateKey = loadPrivateKey(privateString);
  final publicKey = loadPublicKey(publicString);
  assert(privateKey.d != null && publicKey.Q != null);
  final secret = scalarMultiple(
    privateKey.d!,
    publicKey.Q!,
  ); //publicKey.Q * privateKey.d;
  //final secret = publicKey.Q * privateKey.d;
  return secret;
}

/// return a Bytes data secret
List<List<int>> byteSecret(String privateString, String publicString) {
  final secret = rawSecret(privateString, publicString);
  assert(secret.x != null && secret.y != null);
  final xs = secret.x!.toBigInteger()!.toRadixString(16);
  final ys = secret.y!.toBigInteger()!.toRadixString(16);
  final hexX = leftPadding(xs, 64);
  final hexY = leftPadding(ys, 64);
  final secretBytes = Uint8List.fromList(HEX.decode('$hexX$hexY'));
  final pair = [
    secretBytes.getRange(0, 32).toList(),
    secretBytes.getRange(32, 40).toList()
  ];
  //print(secret_bytes);
  //print(pair);
  return pair;
}

/// Encrypt data using target public key
Map pubkeyEncrypt(String privateString, String publicString, String message) {
  convert.Utf8Encoder encoder = convert.Utf8Encoder();
  final enced = pubkeyEncryptRaw(privateString, publicString,
      Uint8List.fromList(encoder.convert(message)));
  //print('enced:${enced["enc"]}');
  return {'enc': convert.base64.encode(enced['enc']), 'iv': enced['iv']};
}

Map pubkeyEncryptRaw(
  String privateString,
  String publicString,
  Uint8List data,
) {
  final secretIV = byteSecret(privateString, publicString);
  final secret = Uint8List.fromList(secretIV[0]);
  final iv = Uint8List.fromList(secretIV[1]);
  //print('s:${secret} iv:${iv}');
  Salsa20Engine _cipher = Salsa20Engine();
  _cipher.reset();
  _cipher.init(true, _buildParams(secret, iv));
  final Uint8List encData = _cipher.process(data);
  return {'enc': encData, 'iv': convert.base64.encode(iv)};
}

/// Decrypt data using self private key
String privateDecrypt(
    String privateString, String publicString, String b64encoded,
    [String b64IV = ""]) {
  Uint8List encdData = convert.base64.decode(b64encoded);
  final rawData =
      privateDecryptRaw(privateString, publicString, encdData, b64IV);
  convert.Utf8Decoder decode = convert.Utf8Decoder();
  return decode.convert(rawData.toList());
}

Uint8List privateDecryptRaw(
    String privateString, String publicString, Uint8List encdData,
    [String b64IV = ""]) {
  final secretIV = byteSecret(privateString, publicString);
  final secret = Uint8List.fromList(secretIV[0]);
  final iv = b64IV.length > 6
      ? convert.base64.decode(b64IV)
      : Uint8List.fromList(secretIV[1]);
  Salsa20Engine _cipher = Salsa20Engine();
  _cipher.reset();
  _cipher.init(false, _buildParams(secret, iv));
  return _cipher.process(encdData);
}

ParametersWithIV<KeyParameter> _buildParams(Uint8List key, Uint8List iv) {
  return ParametersWithIV<KeyParameter>(KeyParameter(key), iv);
}

/// Generate Keypair
AsymmetricKeyPair<PublicKey, PrivateKey> generateKeyPair({
  Uint8List? seed,
}) {
  var keyParams = ECCurve_secp256k1();
  var random = FortunaRandom();
  if (seed == null) {
    random.seed(KeyParameter(_seed(32)));
  } else {
    random.seed(KeyParameter(seed));
  }
  BigInt n = keyParams.n;
  int nBitLength = n.bitLength;
  BigInt d;
  do {
    d = random.nextBigInteger(nBitLength);
  } while (d == BigInt.zero || (d >= n));
  ECPoint q = scalarMultiple(d, keyParams.G);
  return AsymmetricKeyPair(
    ECPublicKey(q, keyParams),
    ECPrivateKey(d, keyParams),
  );
}

Uint8List _seed(length) {
  var random = Random.secure();
  var seed = List<int>.generate(length, (_) => random.nextInt(256));
  return Uint8List.fromList(seed);
}
